### PR TITLE
Input Feature Noise Augmentation: Gaussian noise on all input channels

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

After 5 consecutive failed feature additions this round (chord-camber, Re-walldist, surface normals, arc-length PE, domain-split SRF), evidence strongly suggests the feature engineering space is **saturated** — DSDF + TE coord frame + wake deficit already capture the geometric information the model can use.

The remaining performance gap is primarily in OOD metrics (p_oodc 7.64 vs ensemble 6.6, p_re 6.30 vs ensemble 5.8). This suggests the model **overfits to the exact distribution of input features** in the training set.

**Input Feature Noise Augmentation** adds small Gaussian noise to ALL input feature channels during training. This is a standard regularization technique (Sietsma & Dow 1991, Bishop 1995) that:
1. Smooths the learned function, preventing sharp decision boundaries that break OOD
2. Creates virtual training examples in the neighborhood of each real sample
3. Is equivalent to Tikhonov regularization (L2 penalty on the Jacobian) for small noise magnitudes

The noise is applied AFTER feature normalization so that all channels receive proportional perturbation regardless of their scale.

**Key distinction from other augmentations in baseline:**
- `--aug aoa_perturb`: perturbs ONE specific physical quantity (angle of attack)
- `--aug_dsdf2_sigma`: perturbs DSDF features specifically
- **This**: perturbs ALL input channels uniformly, creating isotropic regularization in feature space

**Key distinction from edward's Re augmentation (#2233):**
- Edward: targets ONE channel (log Re) with a specific physical motivation
- This: targets ALL channels as a general regularization technique

**Confidence:** Medium. Well-established ML technique, but the model already has augmentation on specific channels. The question is whether adding noise to the REMAINING channels (coordinates, boundary IDs, TE frame, wake deficit, etc.) provides additional regularization benefit.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
input_noise: bool = False        # Gaussian noise on all input features during training
input_noise_sigma: float = 0.02  # std of noise (applied after normalization)
```

### Step 2: Apply noise during training

In the **training loop only**, AFTER the input features are assembled and normalized (so all channels are on comparable scales), and BEFORE the forward pass through the model:

```python
if cfg.input_noise and model.training:
    # Add Gaussian noise to all input features
    noise = torch.randn_like(x) * cfg.input_noise_sigma
    x = x + noise
```

**CRITICAL requirements:**
- Apply ONLY during training (NOT during validation, visualization, or re-verification)
- Apply AFTER any input normalization (so noise is proportional to feature scale)
- Apply BEFORE the model forward pass
- The noise should be freshly sampled each batch (different noise per forward pass)

### Step 3: Verify no impact on validation

Double-check that the noise is only applied in the training branch. Validation metrics should be computed on clean (unperturbed) inputs.

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/input-noise-s42" \
  --wandb_group "round18/input-noise-augmentation" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --input_noise --input_noise_sigma 0.02

# Seed 73 — identical but --seed 73 --wandb_name "alphonse/input-noise-s73"
```

### Step 5: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, comparison to baseline, W&B run IDs.

Pay special attention to:
- **p_oodc and p_re**: these are OOD metrics — the target of this regularization
- **p_in**: if this degrades significantly (>3%), noise sigma is too large
- **Training loss curve**: check if the training loss is higher than baseline (expected — noise makes training harder) but validation loss is lower (the regularization signal)

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Summary

Input Feature Noise Augmentation (sigma=0.02 on all channels) **did not improve** over baseline. All 4 metrics regressed substantially — this is the worst regression seen across recent experiments.

### Metrics Table

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | vs Baseline |
|--------|----------|---------|---------|------------|-------------|
| **p_in**  | 11.979 | 12.821 | 13.718 | **13.269** | +1.290 (+10.8%, worse) |
| **p_oodc** | 7.643 | 8.715 | 8.542 | **8.629** | +0.986 (+12.9%, worse) |
| **p_tan** | 28.341 | 30.959 | 29.519 | **30.239** | +1.898 (+6.7%, worse) |
| **p_re**  | 6.300 | 7.120 | 7.273 | **7.197** | +0.897 (+14.2%, worse) |

Metric notation: p_in = val_in_dist mae_surf_p, p_oodc = val_ood_cond mae_surf_p, p_tan = val_tandem_transfer mae_surf_p, p_re = val_ood_re mae_surf_p. Lower is better.

### Run Details

| | Seed 42 | Seed 73 |
|--|---------|---------|
| W&B run | `946piami` | `ej0qyopf` |
| Best epoch | 146 / 500 (wall-clock timeout) | 146 / 500 (wall-clock timeout) |
| Peak VRAM | 46.3 GB | ~46.3 GB |

### Command

```bash
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/input-noise-s42" \
  --wandb_group "round18/input-noise-augmentation" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --input_noise --input_noise_sigma 0.02
# Seed 73 identical with --seed 73 --wandb_name "alphonse/input-noise-s73"
```

### What Happened

Uniform Gaussian noise (sigma=0.02) on all input features after normalization and feature assembly caused **across-the-board regression** of +7-14% on all metrics. This is a clear negative result.

**Why it failed:**

1. **The noise corrupts geometric features that need to be precise.** Input features include normalized (x, y) coordinates, DSDF fields, boundary indicators, TE coordinate frames, and wake deficit features. Many of these encode **exact geometric relationships** (e.g., distance to surface, trailing edge offsets). Adding noise to coordinates directly corrupts the model's spatial understanding — it's equivalent to randomly perturbing the mesh geometry, which physically doesn't make sense.

2. **sigma=0.02 is too large for normalized features.** After normalization, most features have unit variance. Noise with sigma=0.02 adds 2% noise — but for features like binary boundary indicators (is_surface) or precise geometric distances (DSDF), even small perturbations destroy the signal. The existing channel-specific augmentations (AoA perturbation, DSDF sigma) are carefully tuned to only perturb features where physical perturbation is meaningful.

3. **Noise on Fourier PE features is especially harmful.** The 32 Fourier positional encoding features are sine/cosine functions of normalized coordinates. Adding noise to these breaks the learned frequency-position mapping that the Transolver attention depends on. This is analogous to corrupting positional encodings in a Transformer — it destroys the model's ability to do spatial reasoning.

4. **The model already has targeted augmentations.** The baseline uses AoA perturbation, DSDF sigma, gap/stagger sigma, and full DSDF rotation augmentation — all applied to specific channels where physical perturbation is meaningful. Uniform isotropic noise on ALL channels is a blunt instrument that ignores the physical structure of the features.

5. **The OOD metrics (p_oodc, p_re) regressed the MOST** (+12.9%, +14.2%), which is the opposite of what was hoped. This suggests the noise didn't help regularization — it actually made the model's learned representations noisier and less generalizable, not smoother.

### Suggested Follow-ups

- **Channel-selective noise**: Apply noise only to physically-invariant features (the Fourier PE and scalar quantities like curvature) while preserving exact geometric features (coordinates, DSDF, boundary masks). This would require identifying which channels benefit from perturbation.
- **Much smaller sigma**: Try sigma=0.001-0.005 if revisiting this approach — but the fundamental issue is that many input channels encode exact geometry where noise is physically meaningless.
- **Dropout on intermediate representations instead**: Apply noise/dropout INSIDE the model (e.g., on attention scores or hidden states) rather than on the input features. This regularizes the learned function without corrupting the geometric inputs.
- **Consider this a strong negative signal**: 6 consecutive feature/regularization experiments have failed. The model architecture or training procedure may need structural changes (not incremental feature additions) to improve further.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2235 because ADVISOR: Closing — all 4 metrics regressed substantially (p_in +10.8%, p_oodc +12.9%, p_tan +6.7%, p_re +14.2%). Uniform Gaussian noise on all input channels corrupts the precise geometric features (DSDF, coordinates, TE frame, wake deficit) that the model relies on for spatial reasoning. The existing targeted augmentations (AoA perturbation, DSDF sigma, gap/stagger sigma) are already well-tuned for channels where physical perturbation is meaningful. Your analysis is spot-on — noise on exact geometric features is physically meaningless and on Fourier PE is actively destructive. Thank you for the thorough write-up..
